src/core/stream/representation/representation_stream.ts

```diff
@@ -525,27 +525,41 @@ export default function RepresentationStream<T>({
                                              segmentData: evt.value.initializationData,
                                              segmentBuffer });
         return observableMerge(initEncEvt$, pushEvent$);
+      }
 
-      case "parsed-segment":
+      case "parsed-segment": {
         const initSegmentData = initSegmentObject?.initializationData ?? null;
         const { inbandEvents,
                 needsManifestRefresh } = evt.value;
-        const manifestRefresh$ =  needsManifestRefresh === true ?
+
+        // TODO better handle use cases like key rotation by not always grouping
+        // every protection data together? To check.
+        const segmentEncryptionEvent = evt.value.protectionDataUpdate &&
```

Comment:
Is the question mark on the comment or on the need to check `protectionDataUpdate` on a non-init segment?

---

If it is for the comment, it is basically because key rotation may lead to a case where multiple PSSHs boxes with different keyIds may be found in different media segments for the same `Representation` (the only difference being the time at which that segment has been generated / encrypted).

Basically a segment generated at 11.00 would have a PSSH containing the keyId `A` and one generated at 13.54 may have a PSSH containing the keyId `B` because it is encrypted with a different key.

In such cases, we do not want to mix all this Representation's encryption data together (like we're doing here with `representation.getAllEncryptionData()`) because we would be providing multiple keyIds to the `generateRequest` API (both for the old segments and for the newer ones) and the CDM might not know how to react to that.

So for now, key rotation may be broken (but I don't think it ever worked). As it's not the point of this PR, I just added a TODO.

---

If it's why are we checking `protectionDataUpdate` on a non-init segment, this is because those can now both:
  - actually contain initialization data, e.g. when there is only one segment that is both
  - contain a PSSH even if that's only a media segment, as it seems to be how key rotation and [key hierarchy](https://dashif-documents.azurewebsites.net/Guidelines-Security/master/Guidelines-Security.html#CPS-KeyHierarchy) works, as seen in https://dashif-documents.azurewebsites.net/Guidelines-Security/master/Guidelines-Security.html#CPS-cmaf

Reply:
Ok thanks for the explanation. I was wondering at first why the segmentEncryptionEvent was absent from the RepresentationStream, but now i think i have understood !